Use trsm for triangular_solve on CPU

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63570
* #63569
* #63568
* __->__ #63567

The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

It also changes the use of `bool`s to a common enum type to represent
whether a matrix is transposed / conj transposed, etc. This makes the API
consistent, as before, the behaviour of these functions with `transpose=True`
and `conjugate_transpose=True` it was not well defined.
Functions to transform this type into the specific types / chars for the different
libraries are provided under the names `to_blas`, `to_lapack`, `to_magma`, etc.

This is the first of a stack of PRs that aim to improve the performance of 
`linalg.solve_triangular`. `trsm` has an extra parameter (`side`), which allows to 
ellide the copy of the triangular matrix in many cases.

Fixes https://github.com/pytorch/pytorch/issues/56326

Differential Revision: [D30566479](https://our.internmc.facebook.com/intern/diff/D30566479)

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

lezcano committed:
Use trsm for triangular_solve in CPU

This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

The docs are fixed in the next PR of this stack.

Fixes https://github.com/pytorch/pytorch/issues/56326

The current implementation called trtrs for CPU and trsm for CUDA.
See
https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115
for a discussion of the differences between these two functions and why
we prefer trsm vs trtrs on CUDA.

On top of that, this is the first of a stack of PRs that aim to
improve the performance of triangular_solve. trsm has an extra parameter
(`side`), which allows to ellide the copy of the triangular matrix.

[ghstack-poisoned]

lezcano committed:
Update on "Use trsm for triangular_solve in CPU"


This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

The docs are fixed in the next PR of this stack.

Fixes https://github.com/pytorch/pytorch/issues/56326

The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

On top of that, this is the first of a stack of PRs that aim to
improve the performance of triangular_solve. trsm has an extra parameter
(`side`), which allows to ellide the copy of the triangular matrix.

[ghstack-poisoned]

lezcano committed:
Update on "Use trsm for triangular_solve in CPU"


This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

The docs are fixed in the next PR of this stack.

Fixes https://github.com/pytorch/pytorch/issues/56326

The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

On top of that, this is the first of a stack of PRs that aim to
improve the performance of triangular_solve. trsm has an extra parameter
(`side`), which allows to ellide the copy of the triangular matrix.

[ghstack-poisoned]

lezcano committed:
Update on "Use trsm for triangular_solve in CPU"


This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

The docs are fixed in the next PR of this stack.

Fixes https://github.com/pytorch/pytorch/issues/56326

The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

On top of that, this is the first of a stack of PRs that aim to
improve the performance of triangular_solve. trsm has an extra parameter
(`side`), which allows to ellide the copy of the triangular matrix.

[ghstack-poisoned]

lezcano committed:
Update on "Use trsm for triangular_solve in CPU"


This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

The docs are fixed in the next PR of this stack.

Fixes https://github.com/pytorch/pytorch/issues/56326

The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

On top of that, this is the first of a stack of PRs that aim to
improve the performance of triangular_solve. trsm has an extra parameter
(`side`), which allows to ellide the copy of the triangular matrix.

[ghstack-poisoned]

lezcano committed:
Update on "Use trsm for triangular_solve in CPU"


The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

It also changes the use of `bool`s to a common enum type to represent
whether a matrix is transposed / conj transposed, etc. This makes the API
consistent, as before, the behaviour of these functions with `transpose=True`
and `conjugate_transpose=True` it was not well defined.
Functions to transform this type into the specific types / chars for the different
libraries are provided under the names `to_blas`, `to_lapack`, `to_magma`, etc.

This is the first of a stack of PRs that aim to improve the performance of 
`linalg.solve_triangular`. `trsm` has an extra parameter (`side`), which allows to 
ellide the copy of the triangular matrix in many cases.

Fixes https://github.com/pytorch/pytorch/issues/56326

[ghstack-poisoned]

Lezcano commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1950,27 +1957,28 @@ REGISTER_DISPATCH(lu_stub, &apply_lu);
-  magma_int_t n = magma_int_cast(A.size(-1), "A.size(-1)");
-  magma_int_t nrhs = magma_int_cast(b.size(-1), "b.size(-1)");
+  // This allows to pass rectangular A and b when left = True
+  magma_int_t m = magma_int_cast(left ? A.size(-1) : b.size(-2), "m");
+  magma_int_t n = magma_int_cast(b.size(-1), "n");
```
Note that `m` and `n` mean that we are solving a system `AX = B` or `XA = B` with `B` of shape `m x n`. They used to mean something different that was not easy to generalise to systems of the form `XA = B`

lezcano committed:
Update on "Use trsm for triangular_solve in CPU"


The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

It also changes the use of `bool`s to a common enum type to represent
whether a matrix is transposed / conj transposed, etc. This makes the API
consistent, as before, the behaviour of these functions with `transpose=True`
and `conjugate_transpose=True` it was not well defined.
Functions to transform this type into the specific types / chars for the different
libraries are provided under the names `to_blas`, `to_lapack`, `to_magma`, etc.

This is the first of a stack of PRs that aim to improve the performance of 
`linalg.solve_triangular`. `trsm` has an extra parameter (`side`), which allows to 
ellide the copy of the triangular matrix in many cases.

Fixes https://github.com/pytorch/pytorch/issues/56326

[ghstack-poisoned]

IvanYashchuk commented on aten/src/ATen/native/LinearAlgebraUtils.h:
```diff
@@ -12,15 +12,21 @@
 
 namespace at { namespace native {
 
+enum class TransposeType {
```
Could you please add a comment that this enum is used for interfacing with different BLAS-like libraries?

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -10,6 +10,17 @@
+#if AT_BUILD_WITH_LAPACK()
+char to_lapack(TransposeType trans) {
+  switch (trans) {
+    case TransposeType::Transpose: return 't';
+    case TransposeType::NoTranspose: return 'n';
+    case TransposeType::ConjTranspose: return 'c';
```
Probably lowercase letters also work, but all the documentation specifies the uppercase, so let's use that.
```suggestion
    case TransposeType::Transpose: return 'T';
    case TransposeType::NoTranspose: return 'N';
    case TransposeType::ConjTranspose: return 'C';
```

Lezcano replied:
They both work according to BLAS. I just used lowercase because the `to_blas` used in matmul also used them.
Making them uppercase.

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -794,50 +805,46 @@ This is an in-place routine, content of 'B' is overwritten.
-    int* info_working_ptr = &infos_data[i];
-    lapackTriangularSolve<scalar_t>(uplo, trans, diag, n, nrhs, A_working_ptr, lda, B_working_ptr, lda, info_working_ptr);
+    lapackTriangularSolve<scalar_t>(side, uplo, trans, diag, m, n, A_working_ptr, lda, B_working_ptr, ldb);
     // The current behaviour for linear algebra functions to raise an error if something goes wrong
     // or input doesn't satisfy some requirement
     // therefore return early since further computations will be wasted anyway
```
Remove this comment as well.

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -318,20 +318,24 @@ template<> void lapackCholeskyInverse<float>(char uplo, int n, float *a, int lda
 
-template<> void lapackTriangularSolve<c10::complex<double>>(char uplo, char trans, char diag, int n, int nrhs, c10::complex<double> *a, int lda, c10::complex<double> *b, int ldb, int *info) {
-  ztrtrs_(&uplo, &trans, &diag, &n, &nrhs, reinterpret_cast<std::complex<double>*>(a), &lda, reinterpret_cast<std::complex<double>*>(b), &ldb, info);
+template<> void lapackTriangularSolve<c10::complex<double>>(char side, char uplo, char trans, char diag, int n, int nrhs, c10::complex<double> *a, int lda, c10::complex<double> *b, int ldb) {
```
This function is not from LAPACK anymore. `trsm` is from BLAS.
```suggestion
template<> void blasTriangularSolve<c10::complex<double>>(char side, char uplo, char trans, char diag, int n, int nrhs, c10::complex<double> *a, int lda, c10::complex<double> *b, int ldb) {
```

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -10,6 +10,17 @@
 
 namespace {
+#if AT_BUILD_WITH_LAPACK()
+char to_lapack(TransposeType trans) {
```
nit: The same characters are used in BLAS. It should be either `to_blas` or `to_blaslapack`.

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -794,50 +805,46 @@ This is an in-place routine, content of 'B' is overwritten.
 template<typename scalar_t>
-void apply_triangular_solve(Tensor& A, Tensor& B, Tensor& infos, bool upper, bool transpose, bool conjugate_transpose, bool unitriangular) {
+void apply_triangular_solve(Tensor& A, Tensor& B, bool left, bool upper, TransposeType transpose, bool unitriangular) {
 #if !AT_BUILD_WITH_LAPACK()
   TORCH_CHECK(
       false,
       "Calling torch.triangular_solve on a CPU tensor requires compiling ",
       "PyTorch with LAPACK. Please use PyTorch built with LAPACK support.");
 #else
```
With changes in this PR, LAPACK is not needed anymore for this function. It uses BLAS.

IvanYashchuk replied:
`#if !AT_BUILD_WITH_BLAS()` should be used here.

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -794,50 +805,46 @@ This is an in-place routine, content of 'B' is overwritten.
 
-void triangular_solve_kernel(Tensor& A, Tensor& B, Tensor& infos, bool upper, bool transpose, bool conjugate_transpose, bool unitriangular) {
+void triangular_solve_kernel(Tensor& A, Tensor& B, bool left, bool upper, TransposeType transpose, bool unitriangular) {
+  // right solvers are not implemented in LAPACK. The caller always calls with left=True
```
This comment is outdated? `apply_triangular_solve` now calls BLAS, not LAPACK.

IvanYashchuk commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1950,27 +1957,28 @@ REGISTER_DISPATCH(lu_stub, &apply_lu);
-  magma_int_t n = magma_int_cast(A.size(-1), "A.size(-1)");
-  magma_int_t nrhs = magma_int_cast(b.size(-1), "b.size(-1)");
+  // This allows to pass rectangular A and b when left = True
+  magma_int_t m = magma_int_cast(left ? A.size(-1) : b.size(-2), "m");
```
BLAS documentation says that `m` is the number of rows of `B` and `n` is the number of columns of `B`.
https://software.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-c/top/blas-and-sparse-blas-routines/blas-routines/blas-level-3-routines/cblas-trsm.html
Same in MAGMA docs https://icl.cs.utk.edu/projectsfiles/magma/doxygen/group__magma__trsm.html#ga36b2b66d52f384db19cf256cc5e8a4c2
```suggestion
  magma_int_t m = magma_int_cast(b.size(-2), "m");
```

Lezcano replied:
Leaving as it is for now as per the offline discussion.

IvanYashchuk commented on aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu:
```diff
@@ -91,67 +98,65 @@ static void apply_lu_solve_batched_cublas(const Tensor& b, const Tensor& lu, con
-  auto nrhs = cuda_int_cast(B.size(-1), "nrhs");
-  auto lda = std::max<int>(1, m);
+  // This allows to pass rectangular A and B when left = True
+  auto m = cuda_int_cast(left ? A.size(-1) : B.size(-2), "m");
```
```suggestion
  auto m = cuda_int_cast(B.size(-2), "m");
```

Lezcano replied:
Leaving as it is for now as per the offline discussion.

IvanYashchuk commented on aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu:
```diff
@@ -91,67 +98,65 @@ static void apply_lu_solve_batched_cublas(const Tensor& b, const Tensor& lu, con
-  auto nrhs = cuda_int_cast(B.size(-1), "nrhs");
-  auto lda = std::max<int>(1, m);
+  // This allows to pass rectangular A and B when left = True
+  auto m = cuda_int_cast(left ? A.size(-1) : B.size(-2), "m");
```
```suggestion
  auto m = cuda_int_cast(B.size(-2), "m");
```

Lezcano replied:
Leaving as it is for now as per the offline discussion.

IvanYashchuk commented on torch/_torch_docs.py:
```diff
@@ -9691,6 +9691,10 @@ def merge_dicts(*dicts):
 batches of 2D matrices. If the inputs are batches, then returns
 batched outputs `X`
 
+If the diagonal of :attr:`A` contains zeros or elements that are very close to zero and
```
Nice!

IvanYashchuk reviewed (approved):
There are only a few minor things to change. Overall looks great, thank you for overhauling all the transpose chars with proper enums!

lezcano committed:
Update on "Use trsm for triangular_solve in CPU"


The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

It also changes the use of `bool`s to a common enum type to represent
whether a matrix is transposed / conj transposed, etc. This makes the API
consistent, as before, the behaviour of these functions with `transpose=True`
and `conjugate_transpose=True` it was not well defined.
Functions to transform this type into the specific types / chars for the different
libraries are provided under the names `to_blas`, `to_lapack`, `to_magma`, etc.

This is the first of a stack of PRs that aim to improve the performance of 
`linalg.solve_triangular`. `trsm` has an extra parameter (`side`), which allows to 
ellide the copy of the triangular matrix in many cases.

Fixes https://github.com/pytorch/pytorch/issues/56326

[ghstack-poisoned]

Lezcano commented:
Thank you for the review @IvanYashchuk! I just finished addressing the comments.

lezcano committed:
Update on "Use trsm for triangular_solve in CPU"


The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

It also changes the use of `bool`s to a common enum type to represent
whether a matrix is transposed / conj transposed, etc. This makes the API
consistent, as before, the behaviour of these functions with `transpose=True`
and `conjugate_transpose=True` it was not well defined.
Functions to transform this type into the specific types / chars for the different
libraries are provided under the names `to_blas`, `to_lapack`, `to_magma`, etc.

This is the first of a stack of PRs that aim to improve the performance of 
`linalg.solve_triangular`. `trsm` has an extra parameter (`side`), which allows to 
ellide the copy of the triangular matrix in many cases.

Fixes https://github.com/pytorch/pytorch/issues/56326

[ghstack-poisoned]

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.h:
```diff
@@ -37,7 +37,7 @@ template <class scalar_t, class value_t = scalar_t>
 
 template <class scalar_t>
-void lapackTriangularSolve(char uplo, char trans, char diag, int n, int nrhs, scalar_t* a, int lda, scalar_t* b, int ldb, int* info);
+void blasTriangularSolve(char side, char uplo, char trans, char diag, int n, int nrhs, scalar_t* a, int lda, scalar_t* b, int ldb);
```
This is now inside `AT_BUILD_WITH_LAPACK`, we need to move it under `AT_BUILD_WITH_BLAS`.

lezcano committed:
Update on "Use trsm for triangular_solve in CPU"


The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

It also changes the use of `bool`s to a common enum type to represent
whether a matrix is transposed / conj transposed, etc. This makes the API
consistent, as before, the behaviour of these functions with `transpose=True`
and `conjugate_transpose=True` it was not well defined.
Functions to transform this type into the specific types / chars for the different
libraries are provided under the names `to_blas`, `to_lapack`, `to_magma`, etc.

This is the first of a stack of PRs that aim to improve the performance of 
`linalg.solve_triangular`. `trsm` has an extra parameter (`side`), which allows to 
ellide the copy of the triangular matrix in many cases.

Fixes https://github.com/pytorch/pytorch/issues/56326

[ghstack-poisoned]

lezcano committed:
Update on "Use trsm for triangular_solve in CPU"


The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

It also changes the use of `bool`s to a common enum type to represent
whether a matrix is transposed / conj transposed, etc. This makes the API
consistent, as before, the behaviour of these functions with `transpose=True`
and `conjugate_transpose=True` it was not well defined.
Functions to transform this type into the specific types / chars for the different
libraries are provided under the names `to_blas`, `to_lapack`, `to_magma`, etc.

This is the first of a stack of PRs that aim to improve the performance of 
`linalg.solve_triangular`. `trsm` has an extra parameter (`side`), which allows to 
ellide the copy of the triangular matrix in many cases.

Fixes https://github.com/pytorch/pytorch/issues/56326

[ghstack-poisoned]

lezcano committed:
Update on "Use trsm for triangular_solve in CPU"


The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

It also changes the use of `bool`s to a common enum type to represent
whether a matrix is transposed / conj transposed, etc. This makes the API
consistent, as before, the behaviour of these functions with `transpose=True`
and `conjugate_transpose=True` it was not well defined.
Functions to transform this type into the specific types / chars for the different
libraries are provided under the names `to_blas`, `to_lapack`, `to_magma`, etc.

This is the first of a stack of PRs that aim to improve the performance of 
`linalg.solve_triangular`. `trsm` has an extra parameter (`side`), which allows to 
ellide the copy of the triangular matrix in many cases.

Fixes https://github.com/pytorch/pytorch/issues/56326

[ghstack-poisoned]

mruberry reviewed (approved):
Cool work, @Lezcano! And thank you for reviewing, @IvanYashchuk.

I left a couple questions inline but I'll import this now to get it started on the internal tests

lezcano committed:
Update on "Use trsm for triangular_solve in CPU"


The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

It also changes the use of `bool`s to a common enum type to represent
whether a matrix is transposed / conj transposed, etc. This makes the API
consistent, as before, the behaviour of these functions with `transpose=True`
and `conjugate_transpose=True` it was not well defined.
Functions to transform this type into the specific types / chars for the different
libraries are provided under the names `to_blas`, `to_lapack`, `to_magma`, etc.

This is the first of a stack of PRs that aim to improve the performance of 
`linalg.solve_triangular`. `trsm` has an extra parameter (`side`), which allows to 
ellide the copy of the triangular matrix in many cases.

Fixes https://github.com/pytorch/pytorch/issues/56326

Differential Revision: [D30566479](https://our.internmc.facebook.com/intern/diff/D30566479)

[ghstack-poisoned]

lezcano committed:
Update on "Use trsm for triangular_solve in CPU"


The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

It also changes the use of `bool`s to a common enum type to represent
whether a matrix is transposed / conj transposed, etc. This makes the API
consistent, as before, the behaviour of these functions with `transpose=True`
and `conjugate_transpose=True` it was not well defined.
Functions to transform this type into the specific types / chars for the different
libraries are provided under the names `to_blas`, `to_lapack`, `to_magma`, etc.

This is the first of a stack of PRs that aim to improve the performance of 
`linalg.solve_triangular`. `trsm` has an extra parameter (`side`), which allows to 
ellide the copy of the triangular matrix in many cases.

Fixes https://github.com/pytorch/pytorch/issues/56326

Differential Revision: [D30566479](https://our.internmc.facebook.com/intern/diff/D30566479)

[ghstack-poisoned]

mruberry commented:
FYI this needs some internal changes due to the header file changes. I've made them but they need to go through testing

mruberry commented:
Doh! Sorry @Lezcano but my internal changes triggered an internal rebase and now the system can't auto-rebase this because a few of the files have changed. Would you rebase this PR and ping me?

lezcano committed:
Update on "Use trsm for triangular_solve on CPU"


The current implementation called trtrs for CPU and trsm for CUDA.
See https://github.com/pytorch/pytorch/issues/56326#issuecomment-825496115 for a discussion on the differences between
these two functions and why we prefer trsm vs trtrs on CUDA.

This PR also exposes the `side` argument of this function which is used
in the second PR of this stack to optimise the number copies one needs to make
when preparing the arguments to be sent to the backends.

It also changes the use of `bool`s to a common enum type to represent
whether a matrix is transposed / conj transposed, etc. This makes the API
consistent, as before, the behaviour of these functions with `transpose=True`
and `conjugate_transpose=True` it was not well defined.
Functions to transform this type into the specific types / chars for the different
libraries are provided under the names `to_blas`, `to_lapack`, `to_magma`, etc.

This is the first of a stack of PRs that aim to improve the performance of 
`linalg.solve_triangular`. `trsm` has an extra parameter (`side`), which allows to 
ellide the copy of the triangular matrix in many cases.

Fixes https://github.com/pytorch/pytorch/issues/56326

Differential Revision: [D30566479](https://our.internmc.facebook.com/intern/diff/D30566479)

[ghstack-poisoned]

Lezcano commented:
@mruberry rebased!

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30566479).
